### PR TITLE
[feature] Support classical control in `two_qubit_squash`

### PR DESCRIPTION
**Pull request by lmondada**

### CODE REVIEW

File: tket/tests/test_TwoQubitCanonical.cpp [hunk 1]
```diff
@@ -883,6 +883,152 @@ SCENARIO("Test decomposition into 2-CX circuit plus diagonal") {
...
+    circ_0.add_op<unsigned>(OpType::H, {1});
+    circ_0.add_op<unsigned>(OpType::CX, {0, 1});
+
+    Circuit circ_1(2);
+    circ_1.add_op<unsigned>(OpType::ZZPhase, 0.2, {0, 1});
+    exp_circs = {circ_0, circ_1};
+    n_cx = 3;
+    n_tk2 = 2;
+  }
+  GIVEN("A circuit with classical control (3)") {
```
Comment (cqc-alec) [inline tket/tests/test_TwoQubitCanonical.cpp:961]:
```suggestion
  GIVEN("A circuit with classical control (4)") {
```

**Review by cqc-alec — CHANGES_REQUESTED** [review]

File: tket/src/Circuit/include/Circuit/CircUtils.hpp [hunk 2]
```diff
@@ -197,4 +197,26 @@ Circuit with_controls(const Circuit& c, unsigned n_controls = 1);
 std::tuple<Circuit, std::array<Expr, 3>, Circuit> normalise_TK2_angles(
     Expr a, Expr b, Expr c);
 
+using Condition = std::optional<std::pair<std::list<VertPort>, unsigned>>;
```
[the remark this answers is not in the document]
Comment (lmondada) [reply tket/src/Circuit/include/Circuit/CircUtils.hpp:200]: Didn't know! I replaced it with a typedef.

File: tket/src/Transformations/BasicOptimisation.cpp [hunk 3]
```diff
@@ -267,10 +267,20 @@ struct Interaction {
...
 static bool replace_two_qubit_interaction(
-    Circuit &circ, Interaction &i, std::map<Qubit, Edge> &current_edges,
-    VertexList &bin, OpType target, double cx_fidelity, bool allow_swaps) {
+    Circuit &circ, const Interaction &i, std::map<Qubit, Edge> &current_edges,
+    const Condition &cond, VertexList &bin, OpType target, double cx_fidelity,
+    bool allow_swaps) {
   EdgeVec in_edges = {i.e0, i.e1};
   EdgeVec out_edges = {current_edges[i.q0], current_edges[i.q1]};
+  VertexSet vertices = i.vertices;
+  EdgeVec cin_edges, cout_edges;
```
[the remark this answers is not in the document]
Comment (lmondada) [reply tket/src/Transformations/BasicOptimisation.cpp:276]: Good spot, I've removed the duplicate.

File: tket/src/Transformations/BasicOptimisation.cpp [hunk 4]
```diff
@@ -328,18 +346,39 @@ static bool replace_two_qubit_interaction(
...
     bin.insert(bin.end(), sub.verts.begin(), sub.verts.end());
-    circ.substitute(replacement, sub, Circuit::VertexDeletion::No);
+    Circuit replacement_cond(subc.all_qubits(), subc.all_bits());
+    if (cond) {
+      unit_vector_t bits(cond->first.size());
+      for (unsigned i = 0; i < bits.size(); ++i) {
+        bits[i] = Bit(i);
+      }
+      // Add a classically-controlled phase
+      // TODO: use Phase OpType when available
```
[the remark this answers is not in the document]
Comment (lmondada) [reply tket/src/Transformations/BasicOptimisation.cpp:356]: Yes, added.

File: tket/src/Transformations/BasicOptimisation.cpp [hunk 5]
```diff
@@ -328,18 +346,39 @@ static bool replace_two_qubit_interaction(
...
+      replacement_cond.add_conditional_gate(
+          OpType::U1, {2 * replacement.get_phase()},
```
[the remark this answers is not in the document]
Comment (lmondada) [reply tket/src/Transformations/BasicOptimisation.cpp:358]: Done.

File: tket/tests/test_TwoQubitCanonical.cpp (hunk 1 above)
Comment (lmondada) [reply tket/tests/test_TwoQubitCanonical.cpp:961]: Done.

File: tket/src/Circuit/include/Circuit/CircUtils.hpp [hunk 6]
```diff
@@ -197,4 +197,26 @@ Circuit with_controls(const Circuit& c, unsigned n_controls = 1);
...
+ * @param circ the circuit of `v`
+ * @param v the vertex
+ * @return Condition the condition of v
+ */
+Condition get_condition(const Circuit& circ, Vertex v);
+
+/**
+ * @brief Return op or the underlying op if it is an OpType::Conditional.
+ *
+ * Note that this does not support recursive conditional OpTypes.
```
[the remark this answers is not in the document]
Comment (lmondada) [reply tket/src/Circuit/include/Circuit/CircUtils.hpp:218]: Good point. I'm now handling nesting more elegantly.

**Pull request closed without merging**